src/audio/xm64.c

```diff
@@ -0,0 +1,240 @@
+		}
+		assertf(0, "error loading XM64 file -- file corrupted?");
+	}
+
+	uint32_t base_rom_addr = dfs_rom_addr(fn);
+
+	// Count samples
+	int ninst = xm_get_number_of_instruments(player->ctx);
+	int nwaves = 0;
+	for (int i=0;i<ninst;i++)
+		nwaves += xm_get_number_of_samples(player->ctx, i+1);
+
+	// Allocate waveforms (one per XM64's "samples" aka waveforms)
+	player->waves = malloc(sizeof(waveform_t) * nwaves);
+	assert(player->waves);
+	int nw = 0;
+	for (int i=0;i<ninst;i++) {
+		xm_instrument_t *inst = &player->ctx->module.instruments[i];
+		for (int j=0;j<inst->num_samples;j++) {
+			xm_sample_t *samp = &inst->samples[j];
+
+			// Convert offset of samples from relative to absolute
+			samp->data8_offset += base_rom_addr;
+
+			// Initialize the waveform_t structures with information
+			// coming from the XM "sample".
+			samp->wave = &player->waves[nw++];
+			memset(samp->wave, 0, sizeof(waveform_t));
+			samp->wave->name = strdup(fn); // FIXME: maybe better use a proper name here
+			samp->wave->bits = samp->bits;
+			samp->wave->channels = 1;
+			samp->wave->frequency = sample_rate; // fake, will be changed at every key-on
+			samp->wave->len = samp->length;
+			samp->wave->loop_len = samp->loop_type == XM_NO_LOOP ? 0 : samp->loop_length;
+			if (samp->wave->loop_len&1)
+				samp->wave->loop_len -= 1;
+			samp->wave->read = wave_read;
+			samp->wave->ctx = samp;
+		}
+	}
+}
+
+int xm64player_num_channels(xm64player_t *player) {
+	return xm_get_number_of_channels(player->ctx);
+}
+
+void xm64player_play(xm64player_t *player, int first_ch) {
+	assert(first_ch + xm_get_number_of_channels(player->ctx) <= MIXER_MAX_CHANNELS);
+
+	if (!player->playing) {
+		// XM64 header contains the optimal size for sample buffers on each
+		// channel, to minimize memory consumption. To configure it, bump
+		// the frequency of each channel to an unreasonably high value (we don't
+		// know how much we need, so shoot high), but then limit the buffer size
+		// to the optimal value.
+		for (int i=0; i<player->ctx->module.num_channels; i++) {
+			// If the value is 0, the channel is not used. We don't have a way
+			// to convey this (0 would be interpreted as "no limit"), so just
+			// avoid calling the limit function altogether.
+			if (player->ctx->ctx_size_stream_sample_buf[i] != 0)
+				mixer_ch_set_limits(first_ch+i, 0, 1e9, player->ctx->ctx_size_stream_sample_buf[i]);
+		}
+
+		mixer_add_event(0, tick, player);
+		player->first_ch = first_ch;
+		player->playing = true;
+	}
+}
+
+void xm64player_stop(xm64player_t *player) {
+	// Let the mixer callback stop playing
+	player->playing = false;
+}
+
+void xm64player_tell(xm64player_t *player, int *patidx, int *row, float *secs) {
+	// Disable interrupts to try to avoid race conditions with the player
+	// running in a different thread. Notice that this is not sufficient
+	// (you would need some kind of mutex), but let's say good enough,
```

Comment:
technically, the audio thread could have been preempted at any point, eg: while modifying the XM data structures in non-atomic way. Disabling interrupts make sure the current code is not preempted, but it gives you no guaranteed on where exactly the other threads are paused.

In this case, it should work by luck because audio thread is always going to be higher priority so in general it would preempt only on specific points (eg: while waiting for DMA or RSP) so outside the actual XM business logic. Obviously if you call xm64player_tell from an even higher priority thread, all hell break loose.

The only solution here would be to wrap all calls to libxm in a disable/enable preemption, so that all modifications to the XM data structures are kept atomic. I think this is what I will try to do when I get at it.